### ISSUE TITLE
Optimize iterators over mapped bitmaps

### ISSUE THREAD

**Issue by lemire** [lemire/RoaringBitmap#44]
Currently, when iterating over an ImmutableRoaringBitmap, one instantiates object containers and then iterator objects.

We could streamline this process by *not* creating the object containers and iterating directly over the original ByteBuffer.

**Comment by lemire** [comment]
c.c. @gssiyankai

**Comment by gssiyankai** [comment]
@lemire,

:+1:  This will reduce the memory footprint and alleviate garbage collection.

I think we can take the opportunity to give the [UnsafeBuffer](https://github.com/real-logic/Agrona/blob/master/src/main/java/uk/co/real_logic/agrona/concurrent/UnsafeBuffer.java) a try. From my personal experience, accessing ByteBuffer is highly penalized by bounds checking.

**Comment by pwebbitrs** [comment]
I was about to raise this very issue but glad to see it's already on the radar! 

I was just looking at the Byte code generated for  next() on a ImmutableRoaringIntIterator (using mapped bitmaps).   Currently it is being marked as a hotspot on our profiling and it's not hard to see why.  (not a criticism - just an observation - and clearly confirms your concerns above)

```
 0 aload_0
 1 getfield #12 <org/roaringbitmap/buffer/ImmutableRoaringBitmap$ImmutableRoaringIntIterator.iter>
 4 invokeinterface #20 <org/roaringbitmap/ShortIterator.nextAsInt> count 1
 9 aload_0
10 getfield #6 <org/roaringbitmap/buffer/ImmutableRoaringBitmap$ImmutableRoaringIntIterator.hs>
13 ior
14 istore_1
15 aload_0
16 getfield #12 <org/roaringbitmap/buffer/ImmutableRoaringBitmap$ImmutableRoaringIntIterator.iter>
19 invokeinterface #21 <org/roaringbitmap/ShortIterator.hasNext> count 1
24 ifne 40 (+16)
27 aload_0
28 getfield #5 <org/roaringbitmap/buffer/ImmutableRoaringBitmap$ImmutableRoaringIntIterator.cp>
31 invokeinterface #22 <org/roaringbitmap/buffer/MappeableContainerPointer.advance> count 1
36 aload_0
37 invokespecial #7 <org/roaringbitmap/buffer/ImmutableRoaringBitmap$ImmutableRoaringIntIterator.nextContainer>
40 iload_1
41 ireturn
```

All the  `invokeInterface` call's are not ideal in such a critical spot -  same with bounds checking over the buffers - although I'm not sure what you can do here other than (as @lemire  has observed), iterate over the raw bitmap directly.   Keeping GC down is always good from my perspective.   

I really doubt I will have bandwidth to assist here myself but If I find some time, and this is still pending, I'll have a look into it.

**Comment by lemire** [comment]
@pwebbitrs 

I estimate that we could probably make the iterators run twice as fast. I also guess that it requires little coding, just the right approach.

The current approach works... but it has a relatively high overhead.

I should qualify this last statement... we are still better or on par with ImmutableConcise... So we are not bad... it is just that we could be a lot better.

I think it is less a matter of finding time to code it up, and more a matter of finding out *how* to code it up. It is design issue.

I'd be very interested if anyone has a good design...

**Comment by pwebbitrs** [comment]
@lemire  I totally agree on all counts.  And to qualify my comment -  It's certainly not slow :)    It's one of those things one looks at while trying to shave micros/nano's off something that's on the hotpath and get's hit a **lot**. 

Yes it is a bit of a shift in design so yeah it would require more thought than actual code.

**Comment by lemire** [comment]
I think we could probably make things run a lot faster with very simple code using a ``forEach`` idiom instead. It would reduce branching and temporary object creation.

Would a  ``forEach`` idiom meet your needs?

https://docs.oracle.com/javase/8/docs/api/java/lang/Iterable.html#forEach-java.util.function.Consumer-

Strictly speaking, it is a Java 8 feature, and we do not want to require Java 8, but there is not reason it can't be implemented in a way that supports Java 7.

Calling on @gssiyankai  for thoughts on this.

**Comment by pwebbitrs** [comment]
It would depend on a few things.  We are geared up to use the nice "pull" semantics of an iterator, so i suspect a `ForEach` idiom would have a ripple effect in our code base. We'd probably need to wrap it up in an external iterator. So it would depend if all of that produced a net gain once we take into account number of invocations etc.   

If it's a LOT faster then it probably would be a net gain.  It's one of those things that's hard to say without seeing comparative numbers.

**Comment by lemire** [comment]
@pwebbitrs 

I think that the performance gains could be interesting, but if you have to wrap it up, I don't know how, this might defeat the purpose. We don't want users to have to produce boring and possibly  buggy boilerplate code.

Without giving any secret sauce away, can you give you a (simple) example of being "geared up to use the nice pull semantics of an iterator"?

I am looking to explore the design space and see what can be done at the API level.

For example, you could have a ``forEach`` and a `forEachWhileConditionHolds``.

To put it differently, I am willing to do some work to improve the performance with a new API, but only if I know that the resulting API will be useful to people. Otherwise we add yet more code that we have to support for no gain.

**Comment by pwebbitrs** [comment]
By that iterator statement I merely mean an iterator allows you to request next, so you (as the consumer of the data) are effectively _pulling_ the data out of the source.  As I understand the J8 Foreach (and i could be wrong),  it is the ForEach that essentially drives the _pushing_ of data into the consumer.  Also we'd need to then implement Consumers to handle it.    I'm absolutely sure we could make it work - its just a matter of working out the extent of the changes required.  

I just find it a simpler idiom when you want to be able to control the flow of data from one source to another (without passing callbacks, consumers etc).  This also reduces method invocations.  We are one of 'those' users who needs to count tiny stuff like that as it adds up quickly.   So for me, it's often not a matter of whether is something is hard to do (like change to a new idiom),  but more a matter of will it give us a strong net gain - raw performance as well as GC profile etc.   So if it gave some noticeable improvements in those areas I'm sure we could tweak our stuff to take advantage one way or another.  

How easy would it be to do a quick and dirty test of the viability of the ForEach approach?  If some indicative numbers were available without a lot of work put down it could really help to see what kind of improvement we are looking at. 

Many users will be quite comfortable with the ForEach approach. It would be great to hear more opinions on this.  Either way I agree no-one wants to have unused APIs hanging around if they are not useful.

**Comment by lemire** [comment]
@pwebbitrs 

Let us think about what happens in general when you want to iterate over a data structure made of a series of containers.

With an iterator, given how it is currently implemented in this library, you have two layers of iterators. That can't be great. Let us look the code...

https://github.com/RoaringBitmap/RoaringBitmap/blob/master/src/main/java/org/roaringbitmap/RoaringBitmap.java#L73-L81

That's obviously a lot of code, including some temporary memory allocations.

How would the same code work with a `forEach`? It could _almost_ look like this:

```
 forEach(userfunction) {
   for(Container c) {
     c.forEach(userfunction)
   }
 }
```

It is still not perfect because MappeableContainer need to be instantiated, and they have an underlying buffer that also needs to be instantiated. But I think it would be better.

**Comment by pwebbitrs** [comment]
I'm thinking about this now - I'll probably need time to properly consider it in the context of what we are doing - it may even work out  this is actually a better option.   I'd be keen to see how much more performant  this actually is, and how the memory profile is changed.

**Comment by lemire** [comment]
The C port of Roaring (CRoaring) has a forEach called ``roaring_iterate``.
https://github.com/RoaringBitmap/CRoaring/blob/master/include/roaring.h#L119

No benchmark, but the code is really simple.

**Comment by lemire** [comment]
Implemented forEach methods. Working on release.

**Comment by lemire** [comment]
We have ``forEach`` methods as per release 0.6.10.

Though we have no benchmark, I expect that real-life performance could be superior to iterators in some cases. 

It is slightly less intuitive, but it is not really much more code in the end:
https://github.com/RoaringBitmap/RoaringBitmap/blob/master/src/test/java/org/roaringbitmap/TestForEach.java#L14-L22

 @pwebbitrs : I would appreciate your feedback.

c.c. @gssiyankai

**Comment by gssiyankai** [comment]
@lemire 

I've added benchmarks on ``forEach`` with the real datasets we have available.
See 921c8545106378ff920327edbf051b3d1e923ee9
These benchmarks are basically doing the same computation as the ones on ``iterator`` (i.e. sum of the set values).
Below the results. As you can see, the speed-up is impressive ! Well done :+1: 

(dataset)|(immutable)|(type)|forEach (us)|iterate (us)|Gain
------------ | ------------- | ------------ | ------------- | ------------ | -------------
census-income|false|roaring|8372.859|33405.202|-74.94%
census-income|false|roaring_with_run|8299.172|72547.852|-88.56%
census-income|true|roaring|8589.422|28838.484|-70.22%
census-income|true|roaring_with_run|8635.058|67050.811|-87.12%
census-income_srt|false|roaring|7088.752|28260.944|-74.92%
census-income_srt|false|roaring_with_run|2594.446|62868.292|-95.87%
census-income_srt|true|roaring|7119.088|24455.934|-70.89%
census-income_srt|true|roaring_with_run|2727.472|21653.585|-87.40%
census1881|false|roaring|383.961|3542.069|-89.16%
census1881|false|roaring_with_run|363.272|3723.712|-90.24%
census1881|true|roaring|735.543|4947.204|-85.13%
census1881|true|roaring_with_run|675.703|3531.603|-80.87%
census1881_srt|false|roaring|687.784|2654|-74.09%
census1881_srt|false|roaring_with_run|575.316|2782.291|-79.32%
census1881_srt|true|roaring|791.352|3683.164|-78.51%
census1881_srt|true|roaring_with_run|685.377|2685.679|-74.48%
dimension_003|false|roaring|4924.736|18509.042|-73.39%
dimension_003|false|roaring_with_run|1708.753|11744.575|-85.45%
dimension_003|true|roaring|5865.709|18563.199|-68.40%
dimension_003|true|roaring_with_run|3340.176|12661.408|-73.62%
dimension_008|false|roaring|3054.432|12960.621|-76.43%
dimension_008|false|roaring_with_run|1106.715|10305.72|-89.26%
dimension_008|true|roaring|4550.121|13347.178|-65.91%
dimension_008|true|roaring_with_run|1640.206|12209.491|-86.57%
dimension_033|false|roaring|4101.708|18072.517|-77.30%
dimension_033|false|roaring_with_run|1397.527|13866.691|-89.92%
dimension_033|true|roaring|4291.042|16571.449|-74.11%
dimension_033|true|roaring_with_run|1476.69|12864.258|-88.52%
uscensus2000|false|roaring|11.914|39.782|-70.05%
uscensus2000|false|roaring_with_run|13.252|42.944|-69.14%
uscensus2000|true|roaring|59.792|130.512|-54.19%
uscensus2000|true|roaring_with_run|64.909|142.907|-54.58%
weather_sept_85|false|roaring|16332.66|63772.524|-74.39%
weather_sept_85|false|roaring_with_run|16415.56|140083.742|-88.28%
weather_sept_85|true|roaring|17716.504|56896.609|-68.86%
weather_sept_85|true|roaring_with_run|20082.808|69984.113|-71.30%
weather_sept_85_srt|false|roaring|18983.721|72996.774|-73.99%
weather_sept_85_srt|false|roaring_with_run|6530.211|58519.988|-88.84%
weather_sept_85_srt|true|roaring|19132.094|66814.398|-71.37%
weather_sept_85_srt|true|roaring_with_run|7109.982|54346.655|-86.92%
wikileaks-noquotes|false|roaring|123.876|1131.576|-89.05%
wikileaks-noquotes|false|roaring_with_run|619.587|1502.05|-58.75%
wikileaks-noquotes|true|roaring|332.103|845.306|-60.71%
wikileaks-noquotes|true|roaring_with_run|724.61|1273.601|-43.11%
wikileaks-noquotes_srt|false|roaring|306.939|1327.645|-76.88%
wikileaks-noquotes_srt|false|roaring_with_run|202.281|1126.017|-82.04%
wikileaks-noquotes_srt|true|roaring|362.75|1867.466|-80.58%
wikileaks-noquotes_srt|true|roaring_with_run|257.171|1141.4|-77.47%

**Comment by kishoreg** [comment]
Looks great. I missed this thread. What about an api some thing like this

length getNext(int[] array)

it basically fills up the array and the caller gets to pre-create an array
of appropriate size. Method always returns the number of entries that were
written to the array. This should avoid the additional function call we see
in foreach consumer and probably more cpu friendly for the caller to
process a batch of results at a time.

Thoughts?

**Comment by lemire** [comment]
@gssiyankai  I am jealous of your ability to write benchmark and post great looking tables.

I just added a commit to better document ``forEach`` as I think it is a good solution in many cases.

**Comment by lemire** [comment]
@kishoreg 

Yes. The benefit of ``forEach`` is that I was able to implement it in less than one hour. And thanks to @gssiyankai, we know that it is fast.

Your design might be faster in important cases. Maybe. But one would need to implement it and it is maybe not trivial. I don't know.

**Comment by pwebbitrs** [comment]
@lemire  Well done. Very good numbers there.  With that kind of benefit I think we'd be very interested in using this in the ForEach form.  

I think I'm not surprised at the large gains over the existing iterators but I was not expecting it to be quite so good (and consistently good!).  I'll have a closer look at the code and provide more feedback next week.

**Comment by lemire** [comment]
@pwebbitrs Looking forward to your feedback.

You will notice that there is much less code.

**Comment by okrische** [comment]
This is really nice. This almost makes me want to ask for an forEach(IntConsumer,int offset,int length) as well. :>

Is this intendend that it looks more and more like an IntStream as in java8? Does anybody else have the feeling, that roaring bitmap with java8 might return a specialized IntStream via stream(), just like there is a special IntStream for int[] as in Arrays.stream(int[], int startIncl, int startExcl)?

In any way, the IntStream in java has forEachOrdered(IntConsumer), while forEach(IntConsumer) does not warrant the consumption in order.

**Comment by lemire** [comment]
@okrische 

> This almost makes me want to ask for an forEach(IntConsumer,int offset,int length) as well. 

Not hard to implement. Assuming that this is useful to you, and assuming that you are not willing to have a go at implementing it, please open this request as a separate issue so we can track it.

> Is this intendend that it looks more and more like an IntStream as in java8? 

We are not assuming that users support Java 8. With that in mind, I think we do not want to do too much engineering that would become obsolete when we upgrade to Java 8 in a few years.

> Does anybody else have the feeling, that roaring bitmap with java8 might return a specialized IntStream via stream(), just like there is a special IntStream for int[] as in Arrays.stream(int[], int startIncl, int startExcl)?

If you know of a way to do that while still supporting Java 7... 

> In any way, the IntStream in java has forEachOrdered(IntConsumer), while forEach(IntConsumer) does not warrant the consumption in order.

It would be trivial for us to parallelize the forEach call and possibly produce the values out of order.

**Comment by lemire** [comment]
@yichuan1118 I expect ``forEach`` to be generally faster, though the relative benefits might be less for regular bitmaps. Of course, it is easy to tell whether it is the case in the use cases you care about by running a benchmark.
